pkg/backend/snapshot.go

```diff
@@ -49,96 +43,109 @@ type SnapshotPersister interface {
-// by the current plan.
-//
-// Important to note is that, although this SnapshotManager is designed to be easily convertible into a thread-safe
-// implementation, the code as it is today is *not thread safe*. In particular, it is not legal for there to be
-// more than one `SnapshotMutation` active at any point in time. This is because this SnapshotManager invalidates
-// the last persisted snapshot in `BeginSnapshot`. This is designed to match existing behavior and will not
-// be the state of things going forward.
-//
-// The resources stored in the `resources` slice are pointers to resource objects allocated by the engine.
-// This is subtle and a little confusing. The reason for this is that the engine directly mutates resource objects
-// that it creates and expects those mutations to be persisted directly to the snapshot.
+// snapshot, which is immutable and represents the state of the world prior to the application
+// of the current plan, and a journal of operations, which consists of the operations that are being
+// applied on top of the immutable snapshot.
 type SnapshotManager struct {
-	persister      SnapshotPersister    // The persister responsible for invalidating and persisting the snapshot
-	baseSnapshot   *deploy.Snapshot     // The base snapshot for this plan
-	secretsManager secrets.Manager      // The default secrets manager to use
-	resources      []*resource.State    // The list of resources operated upon by this plan
-	operations     []resource.Operation // The set of operations known to be outstanding in this plan
-
-	// The set of resources that have been operated upon already by this plan. These resources could also have
-	// been added to `resources` by other operations but need to be filtered out before writing the snapshot.
-	dones map[*resource.State]bool
+	journal      Journal          // The journal used to record operations performed by this plan
+	baseSnapshot *deploy.Snapshot // The base snapshot for this plan
 
-	completeOps      map[*resource.State]bool // The set of resources that have completed their operation
-	mutationRequests chan<- mutationRequest   // The queue of mutation requests, to be retired serially by the manager
-	cancel           chan bool                // A channel used to request cancellation of any new mutation requests.
-	done             <-chan error             // A channel that sends a single result when the manager has shut down.
-
-	refreshDeletes map[resource.URN]bool // The set of resources that have been deleted by a refresh in this plan.
+	// newResources is a map of resources that have been added to the snapshot in this plan, keyed by the resource
+	// state.  This is used to track the added resources and their operation IDs, allowing us too delete
+	// them later if necessary.
+	newResources       gsync.Map[*resource.State, uint64]
+	operationIDCounter atomic.Uint64 // A counter used to generate unique operation IDs for journal entries.
```

Comment:
If the order doesn't matter, maybe we can add a comment here that the ID does not imply ordering?

Reply:
Yeah it does kind of matter, because of dependencies in the snapshot.  We can send the journal entries in parallel to the service, because the CLI will always finish an operation before it starts working on the dependencies.  However in the service we need the ordering, as we'll have all journal entries there when we want to reconstruct the snapshot, and it might matter for *some* entries that they come after others.  I can try to improve the comment here.